### DISCUSSION TITLE
Can I access the dependant name in a dependancy?

### DISCUSSION THREAD

**Discussion by robteeuwen** [fastapi/fastapi#6503] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, Request, Depends

def func(request: Request):
    # do something that depends on the endpoint that was accessed
    routes = request.app.routes
    path = request.url.path

app = FastAPI(dependencies=[Depends(func)])

@app.get("/")
async def root():
    return {"message": "this is the root"}

@app.get("/someroute")
async def someroute():
    return {"message": "this is some route"}

if __name__ == "__main__":
    uvicorn.run("main:app", host="0.0.0.0", port=8080, reload=True)
```

### Description

I would like to make my dependency `func` behave differently depending on the endpoint that is being accessed. I can figure out which endpoint was called (i.e. the name of the callback function) by matching the path to one of the routes manually. However, since FastAPI has already done the matching, I would prefer to access that matched endpoint from within the dependency, instead of doing the matching again. 

The dependency is called from `solve_dependencies` in `fastapi/dependencies/utils.py`, which contains an object `dependant`, which contains the name of the callback function. Is this accessible from the dependency, for example in a similar way like the request object is accessible? I tried to just do `def func(request: Request, dependant: Dependant)` but that doesn't work. 

Some context: I am writing a dependency to implement RBAC security, which should grant a user access if that user's roles match an externally stored list of required roles. These roles depend on the endpoint. Since I don't want to write a new dependency for each endpoint, the dependency needs to know which endpoint is being called, so it can look up the roles from the list. Manually matching seems a little sketchy and inefficient, since it depends on the order of the routes (if there is more than 1 match) and since FastAPI technically already knows which callback function it will execute after the dependency. 

### Operating System

Linux

### FastAPI Version

0.72.0

### Python Version

3.9.5

**Comment by jgould22** [comment]
Have you thought about doing this in a middleware and not a dependency?

**Comment by lucastosetto** [comment]
@robteeuwen You could use a router for each endpoint, this way this dependency injection will behave exactly like you want. Take a look at: https://fastapi.tiangolo.com/tutorial/bigger-applications/

**Comment by robteeuwen** [comment]
Thanks for your comments. I'm not sure how either of these suggestions would solve my problem though. 

@jgould22, I don't use middleware because the dependency has a subdependency that validates an access token and returns the user's roles. I don't think I can do that in middleware because it can't interact with dependencies. But regardless of that, how would I access the matched endpoint from middleware? 

@lucastosetto, I'm not sure what you mean. My application already uses routers, but I think they face the same problem. Or is there a way to access the matched endpoint from a dependency on a router that doesn't exist in dependencies on an entire app object?

**Comment by robteeuwen** [comment, marked as answer]
I noticed that the `__call__` method of the `Router` object (in starlette/routing.py) handles the matching, and stores the callback function of the matched endpoint in the ASGI connection scope (`scope`), which is attached to the request object that the dependency also has access to. So accessing the name of the callback function that will be executed is as simple as `request.scope['endpoint'].__name__`. Hope it helps someone.
